src/hooks/post_parse.c

```diff
@@ -0,0 +1,146 @@
+    context.oidList = oidList;
+    return operator_used_walker(node, &context);
+}
+
+typedef struct
+{
+    List *sortGroupRefs;
+} SortGroupRefContext;
+
+static bool sort_group_ref_walker(Node *node, SortGroupRefContext *context)
+{
+    if(node == NULL) return false;
+    if(IsA(node, Query)) {
+        Query    *query = (Query *)node;
+        ListCell *lc;
+        foreach(lc, query->sortClause) {
+            SortGroupClause *sortGroupClause = (SortGroupClause *)lfirst(lc);
+            context->sortGroupRefs = lappend_int(context->sortGroupRefs, sortGroupClause->tleSortGroupRef);
+        }
+        return query_tree_walker((Query *)node, sort_group_ref_walker, (void *)context, 0);
+    }
+    return expression_tree_walker(node, sort_group_ref_walker, (void *)context);
+}
+
+static List *get_sort_group_refs(Node *node)
+{
+    SortGroupRefContext context;
+    context.sortGroupRefs = NIL;
+    sort_group_ref_walker(node, &context);
+    return context.sortGroupRefs;
+}
+
+typedef struct
+{
+    List *oidList;
+    List *sortGroupRefs;
+    bool  usedCorrectly;
+} OperatorUsedCorrectlyContext;
+
+static bool operator_used_correctly_walker(Node *node, OperatorUsedCorrectlyContext *context)
+{
+    if(node == NULL) return false;
+    if(IsA(node, Query)) return query_tree_walker((Query *)node, operator_used_correctly_walker, (void *)context, 0);
+    if(IsA(node, TargetEntry)) {
+        TargetEntry *te = (TargetEntry *)node;
+        if(te->resjunk && list_member_int(context->sortGroupRefs, te->ressortgroupref)) {
+            if(IsA(te->expr, OpExpr)) {
+                OpExpr *opExpr = (OpExpr *)te->expr;
+                if(list_member_oid(context->oidList, opExpr->opno)) {
+                    Node *arg1 = (Node *)linitial(opExpr->args);
+                    Node *arg2 = (Node *)lsecond(opExpr->args);
+                    bool  isVar1 = IsA(arg1, Var);
+                    bool  isVar2 = IsA(arg2, Var);
+                    if(isVar1 && isVar2) {
+                        return false;
+                    } else if(!isVar1 && !isVar2) {
+                        return true;
+                    } else if(isVar1) {
+                        return operator_used_correctly_walker(arg2, context);
+                    } else {
+                        return operator_used_correctly_walker(arg1, context);
+                    }
+                }
+            }
+        }
+    }
+    if(IsA(node, OpExpr)) {
+        OpExpr *opExpr = (OpExpr *)node;
+        if(list_member_oid(context->oidList, opExpr->opno)) {
+            return true;
+        }
+    }
+
+    return expression_tree_walker(node, operator_used_correctly_walker, (void *)context);
+}
+
+static bool is_operator_used_correctly(Node *node, List *oidList, List *sortGroupRefs)
+{
+    OperatorUsedCorrectlyContext context;
+    context.oidList = oidList;
+    context.sortGroupRefs = sortGroupRefs;
+    return !operator_used_correctly_walker(node, &context);
```

Comment:
This function has very strange semantics:

'is_operator_used_correctly' returns the opposite of 'operator_used_correctly_walker' ?

Reply:
Walkers return false in general. They return true when you exit prematurely. In my case, I want to exit prematurely when something is wrong, i.e., the operator is used incorrectly.

I could rename everything to "incorrectly"

Reply:
Renamed.